src/modal/modal-ref.ts

```diff
@@ -106,22 +126,32 @@ export class NgbModalRef {
   }
 
   private _removeModalElements() {
-    const windowNativeEl = this._windowCmptRef.location.nativeElement;
-    windowNativeEl.parentNode.removeChild(windowNativeEl);
-    this._windowCmptRef.destroy();
+    // hiding window
+    this._windowCmptRef.instance.hide().subscribe(() => {
+      const windowNativeEl = this._windowCmptRef.location.nativeElement;
+      windowNativeEl.parentNode.removeChild(windowNativeEl);
+      this._windowCmptRef.destroy();
+
+      if (this._contentRef && this._contentRef.viewRef) {
+        this._contentRef.viewRef.destroy();
+      }
 
-    if (this._backdropCmptRef) {
-      const backdropNativeEl = this._backdropCmptRef.location.nativeElement;
-      backdropNativeEl.parentNode.removeChild(backdropNativeEl);
-      this._backdropCmptRef.destroy();
-    }
+      this._windowCmptRef = <any>null;
+      this._contentRef = <any>null;
+      this._closed.next();
+      this._closed.complete();
```

Comment:
I was thinking about this too, but came to the conclusion, that we don't care about backdrop, because it's purely cosmetic. There is no key/mouse handling there or anything. Also wanted to avoid overcomplicating things here, because we have to `zip` + backdrop can be optional, etc.

> Suppose user override the fading transition for the backdrop to be 4s

As for this, I don't see why anybody would do this in real life to be honest :) it will be a matter of milliseconds and users will care about window itself, so I can't see how the app will be non-usable from the JS point of view. Otherwise - sure, while backdrop is there - you can't click anywhere, but this has nothing to do with JS logic to me.

But I can change this of course!
@fbasso do you have any input on this?

Reply:
Good remark done by @benouat, it's a question we need to wonder, but I'm more in favor of keeping things simple. If no customization is done, the users don't see the difference, and if an issue is raised later, it will still be time to fix it

Reply:
> that we don't care about backdrop, because it's purely cosmetic

Actually I think we should care about it and I disagree with that. In my example, as soon as `closed` is fired, the app might trigger some UI updates to display things, and if backdrop is still transitioning, these changes will occur below the backdrop, which is bad.

I am convinced that `closed` should only emit after backdrop is gone (in the case there is a backdrop obviously)

Reply:
Ok, will do the change for consistency. I was 50/50 on this one.